### ISSUE TITLE
Logging: Provide a way to capture/redirect internal logs

### ISSUE THREAD

**Issue by anakrish** [openenclave/openenclave#3073]
OE SDK currently provides a way for users to turn on internal logging for debug-enabled enclaves by setting the **OE_LOG_LEVEL** and **OE_LOG_DEVICE** environment variables.

There is also support for custom-logging format: [LoggingFormat.md](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/LoggingFormat.md)

Three separate teams/users have expressed the need to be able to capture and redirect the internal logs generated by OE SDK:

1."Some applications will have their own mechanism for communication between enclave and host, and will want the log messages to use that/piggy back on existing traffic (we are in that situation), so something that would be beneficial is a logging mechanism that clearly separates message generation from their delivery, and allows user-provided delivery." @achamayou  (CCF)

2."This is mainly for enabling diagnosing issues from ACC customers running apps using OE SDK.Ideally, we want a want to **redirect all the existing logging** (hopefully we could expand it) via a plug-in so that it could redirect it to Kusto." @soccerGB (ACC)

3."If you just let me see your logging strings raw, I can take it from there. We just need to have a way for OE to take a c**allback for logging.**" - Internal Microsoft team

**The ask is specifically to be able to capture/redirect internal logs**. This is NOT an ask to add Azure specific code to the OE SDK, or to add telemetry.

**Comment by anakrish** [comment]
One user remarked that just providing the ability to install a `logging-callback` would satisfy their needs:

```c
Typedef void(oe_logging_callback*)(void *context, char *message);

oe_result oe_set_logging_callback(void *context, oe_logging_callback* callback);

void oe_set_logging_callback(void *context, oe_logging_callback *callback)
{
    If (oe_mutex_lock(&_log_lock) == OE_OK)
    {
        _oe_logging_callback = callback;
        _oe_logging_context = context;
        oe_mutex_unlock(&_log_lock);
    }
}
```
"Then in oe_log_message, somewhere around line 347 or so, add:"

```c
if (!_oe_logging_callback)
{
    _oe_logging_callback(_oe_logging_context, message);
}
```

"I think that’s all that needs to happen to satisfy our requirements."

**Comment by anakrish** [comment]
Though @soccerGB used the word "plug-in" in the issue @2895, he clarified that it could be something as simple as providing a custom Ocall implementation for the ocalls that print the messages:

@anakrish: "A plugin, as you point out,  would be one way to do it (though the term plugin can encompass anything from a very formal method to just installing a hook for "printing" the messages."

@soccerGB "The plug-in that I was referring to could be done through replacing Ocall, which could become a general mechanism for substituting certain OCall implementations. "

**Comment by jhand2** [comment]
Dup of #2895? We should only keep one or the other open

**Comment by anakrish** [comment]
To clarify, TEEs (SGX attestation libraries, OPTEE base libraries) also generates logs. This issue does not require that those logs also be captured via the internal logging mechanism.

**Comment by soccerGB** [comment]
"One user remarked that just providing the ability to install a logging-callback would satisfy their needs:"
My original request is more than above. I prefer to see this logging redirecting mechanism could be enable from outside of an enclave app, which would make it easier to redirect to anywhere without the limitation of C/C++ language.

**Comment by anakrish** [comment]
@soccerGB Do you have a design in mind? I think that once we have the ability to redirect logs from within the enclave app, then the other two users' needs can be satisfied.
Can we build this ability to `enable from outside of an enclave app` atop the ability to enable it from within an enclave app?

**Comment by anakrish** [comment]
First scenario is addressed by #3342

**Comment by letmaik** [comment]
How does #3342 address the first requirement of enabling it from inside the enclave? As far as I see the log callback can currently only by set from the host.
